plutus-core/testlib/PlutusCore/Generators/PIR/Substitutions.hs

```diff
@@ -0,0 +1,312 @@
+-- editorconfig-checker-disable-file
+
+module PlutusCore.Generators.PIR.Substitutions where
+
+import Control.Monad.Except
+
+import Control.Monad.Reader
+import Control.Monad.State.Strict
+import Data.Map.Strict (Map)
+import Data.Map.Strict.Internal qualified as Map
+import Data.Maybe
+import Data.MultiSet (toMap)
+import Data.MultiSet.Lens
+import Data.Set (Set)
+import Data.Set qualified as Set
+import Data.Set.Lens
+import GHC.Stack
+import Test.QuickCheck hiding (reason)
+
+import PlutusCore.Default
+import PlutusCore.Name
+import PlutusCore.Pretty
+import PlutusIR
+import PlutusIR.Subst
+
+import PlutusCore.Generators.PIR.Common
+import PlutusCore.Generators.PIR.GenerateTypes
+import PlutusCore.Generators.PIR.ShrinkTypes
+import PlutusCore.Generators.PIR.Utils
+
+type TypeSub = Map TyName (Type TyName DefaultUni ())
+
+unificationFailure :: (MonadError String m, Pretty a, Pretty b) => a -> b -> m any
+unificationFailure x y = throwError $ concat
+    [ "Failed to unify\n\n  "
+    , display x
+    , "\n\nand\n\n  "
+    , display y
+    , "\n\n"
+    ]
```

Comment:
Why not use `vcat` to build the combined doc and then `display` it at the end? Simpler than inserting all those newlines manually?

Reply:
I want a specific number of newlines and I'd rather write them out manually than go fishing in the docs of `prettyprinter` in order to make them `hardline`s and trying to understand if it's going to collapse them or something.